### PR TITLE
RFC-0016: Masked reductions and normalizations

### PR DESCRIPTION
**Pull request by cpuhrsch**
This RFC Discusses semantics and implementation details of masked reduction and normalization operators.

[Rendered](https://github.com/pytorch/rfcs/blob/8cb9ce7fe84724099138dc281080c74ad1bc2cca/RFC-0016-Masked-reductions-and-normalizations.md)

### CODE REVIEW

**Comment by ezyang** [comment]
> does this mean you'd prefer to see this released and packaged out of tree first before considering inclusion in the core?

Not necessarily; I'm referring to this part of the spec:

> Indeed the best way to describe the behavior is to implement it. Please note that this is only meant to describe semantics and is not an actual implementation.

wouldn't be a long step to have an executable specification that people can play around with.

**Comment by mruberry** [comment]
Since the nan* reductions, like [nansum](https://pytorch.org/docs/master/generated/torch.nansum.html?highlight=nan#torch.nansum), are existing masked reductions we should be sure the semantics are equivalent. This proposal just allows the mask to be specified directly rather than by value. Supporting more general value-based masking might be interesting in the future, too.

cc @heitorschueroff

**Comment by cpuhrsch** [comment]
@ezyang - agreed, I'm wondering whether or when we should create an out-of-tree Python-only prototype for a MaskedTensor.

@mruberry - you can always get a value-based mask by e.g. ```masked_sum(input, ~(input != 4))``` or ```masked_sum(input, ~(input != input))``` for nan.

**Comment by pearu** [comment]
> `masked_sum(input, ~(input != input))` for nan.

Nit: `masked_sum(input, input == input)` would work for the nan case as well.

**Comment by ezyang** [comment]
> agreed, I'm wondering whether or when we should create an out-of-tree Python-only prototype for a MaskedTensor.

If it's just one person, probably sticking it in a colab is good enough. Multiple people wanting to work on the semantics ~> put it in GitHub somewhere.

File: RFC-0016-Masked-reductions-and-normalizations.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+* softmax
+* log_softmax
+* norm / native_norm
+
+Sparse CSR: None
+
+**Operator constraints and general signature - Reductions**
+
+**Input types:** A mask is a boolean tensor. It accompanies a dense Tensor of the same shape. If an entry is True the corresponding element at the same index in the paired dense Tensor is a "valid" value. If it is False it is not. "valid" here means that this value is meant to be included in the computation and otherwise is meant to be ignored. This matches the semantics of masked_scatter, masked_select and masked_fill.
```
Comment (pearu) [inline RFC-0016-Masked-reductions-and-normalizations.md:108]: This definition of mask mismatches the definition of [numpy.ma mask](https://numpy.org/doc/stable/reference/maskedarray.generic.html#what-is-a-masked-array):
```
When an element of the mask is False, the corresponding element of the associated array is valid
and is said to be unmasked. When an element of the mask is True, the corresponding element of
the associated array is said to be masked (invalid).
```
Possible solutions:
1. Adjust the mask definition to match the one in numpy.ma
2. Rename "mask" to "valid"
3. Document the mismatch with numpy.ma

I would vote for 1 because the mismatch with the three non-arithmetic ops `masked_scatter`, ... would not be as bad as the mismatch with arithmetic ops of more widely used numpy.ma module, IMHO.

**Review by pearu — COMMENTED** [review]
I have two suggestions
- re mask definition that mismatches with the one from numpy.ma
- re API of masked operations to match the requirements of ReductionOpInfo

File: RFC-0016-Masked-reductions-and-normalizations.md (hunk 1 above)
Comment (pearu) [reply RFC-0016-Masked-reductions-and-normalizations.md:108]: I will redraw my vote for option 1 and suggest defining:
```
A mask is a boolean tensor of selection.
```

Comment (cpuhrsch) [reply RFC-0016-Masked-reductions-and-normalizations.md:108]: I think the current definition of mask adheres to ```A mask is a boolean tensor of selection.``` if I understand you correctly?

Interestingly enough our [MHA implementation](https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention) uses True to indicate an element is meant to be ignored.

But like you mention the other ops use the inverse of this definition.

Comment (pearu) [reply RFC-0016-Masked-reductions-and-normalizations.md:108]: > I think the current definition of mask adheres to `A mask is a boolean tensor of selection.` if I understand you correctly?

yes

> Interestingly enough our [MHA implementation](https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention) uses True to indicate an element is meant to be ignored.

I guess it boils down to not assuming that "mask" is a uniquely defined concept and the exact meaning of mask must be documented in functions docstrings.

> But like you mention the other ops use the inverse of this definition.

In addition, https://numpy.org/doc/stable/reference/generated/numpy.sum.html uses `where` argument as a "mask of selection".

File: RFC-0016-Masked-reductions-and-normalizations.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+**Operator coverage:** As a first step we should implement the basics: sum, prod, mean, amin, amax.
+
+**Semantics:** Indeed the best way to describe the behavior is to implement it. Please note that this is only meant to describe semantics and is not an actual implementation. It also implies the gradients for these operations. Whether an input or mask is backed by a dense or compressed storage layout has no influence on the semantics of these operations. More concretely, unspecified values for sparse layouts are always zero in this context. If the mask is None the behavior of the operator is equivalent to the unmasked variant.
+
+**Input layouts:** If an input is passed a sparse Tensor, it will return a sparse output. At first the layout of the mask and input must match. If the input is a sparse Tensor, the layout of the mask must be too. This is to ease kernel development and might quickly change, if there is user demand. However, the user can always convert a mask to sparse or dense before invocation.
+
+
+```
+def masked_sum(input, dim, keepdim, dtype, mask):
+    return torch.sum(input * mask, dim, keepdim, dtype=dtype)
```
Comment (cpuhrsch) [inline RFC-0016-Masked-reductions-and-normalizations.md:121]: NOTE: This actually requires masked_fill, because we have no guarantee that any masked values of input are valid.
```
>>> torch.tensor([float('inf')]) * torch.tensor([False])
tensor([nan])
>>> torch.tensor([float('inf')]).masked_fill(~torch.tensor([False]), 0)
tensor([0.])
```

File: RFC-0016-Masked-reductions-and-normalizations.md (hunk 1 above)
Comment (cpuhrsch) [reply RFC-0016-Masked-reductions-and-normalizations.md:108]: Based on our offline discussion with Ralf I think it's safe to go for mask semantics that adhere to those described in this RFC.
